### PR TITLE
Nix integration

### PR DESCRIPTION
**Pull request by ttuegel**
These commits add Nix integration to `cabal-install`. If the global `--enable-nix` option is specified and a `shell.nix` file exists, `cabal` will run the affected commands in a `nix-shell` instance.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Nice!

Needs a note in the changelog, though. Will `new-build`be supported as well?

**Comment by k0001** [comment]
From `nix-shell`'s man page:

> `nix-shell` defaults to `shell.nix` if it exists, and `default.nix` otherwise.

Shouldn't `cabal` try and do the same so that the behavior of `cabal` is more predictable to `nix-shell` users?

**Comment by ttuegel** [comment]
> Also, would it be possible to set up some kind of CI for this stuff?

@23Skidoo Maybe? I'm not sure what you have in mind, could you elaborate?

**Comment by ttuegel** [comment]
> Shouldn't cabal try and do the same so that the behavior of cabal is more predictable to nix-shell users?

@k0001 Well, I thought about that. The fact that `nix-shell` respects `default.nix` is mostly a historical accident. Once there were no `shell.nix` files, but more and more packages need different expressions for building versus developing. This distinction is usually represented by having different `shell.nix` and `default.nix` files. In fact, the need to distinguish was mostly driven by Haskell packages! I imagine most Nix users will set `nix: True` in their `~/.cabal/config` and I don't want them to get into trouble if a package doesn't provide `shell.nix` and its `default.nix` isn't suitable for `nix-shell`. But as you pointed out, Nix users may also be confused when they encounter a `default.nix` that *is* suitable for `nix-shell`, but gets ignored. What do you think?

File: Cabal/doc/nix-integration.rst [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+- ``cabal freeze``
+- ``cabal gen-bounds``
+- ``cabal run``
+
+If the package does not provide a ``shell.nix``, ``cabal`` runs normally.
+
+Creating Nix Expressions
+------------------------
+
+The Nix package manager uses a functional language to describe packages, known as expressions. The fastest way to create a Nix expression for a Cabal package is with the ``cabal2nix`` tool. To create a ``shell.nix`` expression for the package in the current directory, run this command:
```
Comment (ttuegel) [inline Cabal/doc/nix-integration.rst]: @grayjay That's not quite what I meant. Does the updated version read better?

Comment (grayjay) [inline Cabal/doc/nix-integration.rst]: Yes, I think that's clearer.

**Comment by k0001** [comment]
I find it useful to define the behavior both for nix-shell and nix-build in
a single default.nix file. But  anyway, seeing as this is the default
behavior for nix-shell, I think supporting it in cabal as well would lead
to more predictable results. At worst, if your default.nix is not nix-shell
aware, you'll get some kind of error in the command line right away.

**Comment by ttuegel** [comment]
> At worst, if your default.nix is not nix-shell aware, you'll get some kind of error in the command line right away.

I can live with that!

**Comment by ttuegel** [comment]
@23Skidoo I'm getting timeouts on Travis. What is the etiquette for merging in that case?

**Comment by ezyang** [comment]
It doesn't look like you made any changes that should have made the build go slower; the OS X machines are just inexplicably 20-50% slower. Unfortunately, I don't see any reports on Travis CI.  I went ahead and restarted the builds in hopes that it was just a temporary slowdown. If not, we'll have to play some games to get it down to acceptable level (maybe finally make cabal-install-lib so we don't have to build the cabal-install libraries three times to test them!)

**Comment by ezyang** [comment]
Fixed!

**Comment by cstrahan** [comment]
Hey, this looks pretty great!

One thing I'm curious about: I'm almost always in a Nix shell, so I imagine I'd have to either open a fresh shell or invoke cabal as `env -u IN_NIX_SHELL cabal ...`. Assuming my reasoning is correct, might it make more sense to use some other heuristic - perhaps an env var specifically for this purpose (e.g. `CABAL_IN_NIX_SHELL`)?

**Comment by ttuegel** [comment]
> perhaps an env var specifically for this purpose (e.g. CABAL_IN_NIX_SHELL)?

That's a good idea. `cabal` should not care if it is initially executed inside a Nix shell; we only want to distinguish between "inside a shell started by cabal" and not.

**Pull request merged**